RT-93 message packer generation

ismail-s opened the pull request:
- Throw error when missing tag when generating messages
- Improve logging around message generation
- Add more unittests around message generation
- Add a `messageId` optional query param that can be used to specify the message id so as to use the same message id in the message header and HL7 body.
  - Add an integration test to test this (in addition to unittest)
- Fix (in e3b8a80) a bug in configuration manager (which I hit with this PR) where the same config object was being returned instead of a copy (so any mutation to the config object would get shared).

gareth-allan committed:
RT-28: Initial package structure

gareth-allan committed:
RT-28: Added proposed modules/classes

gareth-allan committed:
Refactored HttpTransport to OutboundTransmission

gareth-allan committed:
Merge branch 'develop' into feature/RT-28-application-refactor

gareth-allan committed:
RT-28: Correct typo in an OutboundTransmission docstring

gareth-allan committed:
RT-28: Refactored InteractionsConfigFile to ConfigurationManager

gareth-allan committed:
RT-28: Refactored routing component into new structure

gareth-allan committed:
RT-28: Removed common.logging package

gareth-allan committed:
RT-28: Added message building functionality to ebXML envelopes

gareth-allan committed:
RT-28: Added message parsing functionality to ebXML envelopes

gareth-allan committed:
RT-28: Extracted a common superclass for envelope tests

gareth-allan committed:
RT-28: Added inbound request handler

gareth-allan committed:
Merge branch 'develop' into feature/RT-28-application-refactor

gareth-allan committed:
RT-28: Added sync-asnc workflow implementation

gareth-allan committed:
RT-28: Fixed import format in EbxmlEnvelope

gareth-allan committed:
RT-28: Fixed InboundHandler's test class name

gareth-allan committed:
RT-28: Fixed CommonOutbound's class name

gareth-allan committed:
RT-28: Added synchronous outbound request handler implementation

gareth-allan committed:
RT-28: Fixed imports in inbound handler tests

gareth-allan committed:
RT-28: Ported integration tests to refactored application

gareth-allan committed:
RT-28: Ported main.py to new application structure

gareth-allan committed:
RT-28: Updated pipeline to point to new app

gareth-allan committed:
RT-28: Added .coveragerc to new app

gareth-allan committed:
RT-28: Removed mhs-reference-implementation directory

gareth-allan committed:
RT-28: Update merge from develop

ismail-s committed:
RT-93 raise error on missing tag

ismail-s committed:
RT-93 log when missing tag

ismail-s committed:
RT-93 change log message to use new logging approach

ismail-s committed:
RT-93 add more tests for message generation

ismail-s committed:
RT-93 allow passing message id in

ismail-s committed:
RT-93 split line

ismail-s committed:
Merge branch 'develop' into feature/message-packer-generation

ismail-s committed:
RT-93 fix routing

ismail-s committed:
RT-93 fix bad merge

ismail-s commented on common/builder/pystache_message_builder.py:
```diff
@@ -20,4 +26,16 @@ def build_message(self, message_dictionary):
+            raise MessageGenerationSOAPFaultError()
+
+
+class MessageGenerationSOAPFaultError(Exception):
```
Thinking of changing this to something like `MessageGenerationError` or some more specific thing like `MessageGenerationMissingTagError`. The reason I bunged `SOAPFault` in the name originally was because it was mentioned in [Confluence](https://gpitbjss.atlassian.net/wiki/spaces/RTDel/pages/1562149077/ebXML+SOAP+MessageHeader+Ack+to+Spine?atlOrigin=eyJpIjoiODg4MDEwOTk3MDYxNGI2Y2FmZmU5YzBmNjRjMjIwOTkiLCJwIjoiYyJ9).

Philip-Woods replied:
Might suffice to use `MessageGenerationError` and pass in a string specifying it's due to a missing tag?

ismail-s committed:
RT-93 add integration test for passing in message id

Philip-Woods commented on common/builder/pystache_message_builder.py:
```diff
@@ -1,6 +1,11 @@
 from pystache import Renderer
+from pystache.common import MissingTags
+from pystache.context import KeyNotFoundError
+from utilities.integration_adaptors_logger import IntegrationAdaptorsLogger
```
Don't know if we're following/enforcing this so probably optional, but the google style guide says to avoid importing specific classes in favor of the module

Philip-Woods commented on common/builder/tests/test_pystache_message_builder.py:
```diff
@@ -1,7 +1,7 @@
 from unittest import TestCase
 
-from builder.pystache_message_builder import PystacheMessageBuilder
+from builder.pystache_message_builder import PystacheMessageBuilder, MessageGenerationSOAPFaultError
```
(Maybe optional) Again modules over classes

Philip-Woods commented on common/builder/tests/test_pystache_message_builder.py:
```diff
@@ -19,3 +19,7 @@ def test_build_message(self):
+
+    def test_build_message_errors_on_missing_tag(self):
+        with self.assertRaises(MessageGenerationSOAPFaultError):
+            self.builder.build_message({})
```
As a part of the Gp summary upload there are optional sections, just to check that we want that behavior where this throws when an optional tag isn't present?

ismail-s replied:
If I'm thinking about this correctly:
If we want the optional tags to default to False then it seems that this would have to be done at a point where it is known what the template is (so that we know what the optional tags are), and then set defaults for those tags specifically. I haven't done this, I think we only need to do this if we find that there's a decent chunk of time we're spending specifying these default tags.

Philip-Woods replied:
Seems reasonable to leave it as it is, like you say we can always rethink it if we're defaulting a lot of tags

Philip-Woods commented on mhs/mhs/common/messages/ebxml_envelope.py:
```diff
@@ -55,11 +56,14 @@ def serialize(self) -> typing.Tuple[str, str]:
         timestamp = message_utilities.MessageUtilities.get_timestamp()
         ebxml_message_dictionary[TIMESTAMP] = timestamp
-        logging.debug("Creating ebXML message with message ID '%s' and timestamp '%s'", message_id, timestamp)
+        logger.info('0001', 'Creating ebXML message with {MessageId} and {Timestamp}',
```
Just an idea but is it worth doing logging codes in the hundreds per method/ series of activities, this way when we add/remove a log call we don't have to update the whole file to shift the values.

ismail-s replied:
This feels like something to discuss with @dshipp-bjss.

Philip-Woods commented on mhs/mhs/common/workflow/sync_async.py:
```diff
@@ -33,18 +33,24 @@ def __init__(self, config_manager: configuration_manager.ConfigurationManager,
 
-    def prepare_message(self, interaction_name: str, content: str) -> typing.Tuple[bool, str, str]:
+    def prepare_message(self, interaction_name: str, content: str, message_id: typing.Optional[str] = None) \
+            -> typing.Tuple[bool, str, str]:
```
As I was saying earlier google suggests importing modules not classes, however there is an exemption for the typing  [module](http://google.github.io/styleguide/pyguide.html#22-imports)

Philip-Woods commented on mhs/mhs/common/workflow/sync_async.py:
```diff
@@ -33,18 +33,24 @@ def __init__(self, config_manager: configuration_manager.ConfigurationManager,
         interaction_details = self._get_interaction_details(interaction_name)
 
+        if message_id is not None:
+            interaction_details[ebxml_envelope.MESSAGE_ID] = message_id
```
This will apply if `message_id` is an empty string, would there ever be a situation where we want an empty message id or should we just not have the pair?

ismail-s replied:
I doubt we'd want an empty message id, but I don't think we should do any validation here as this isn't the interface we expose to the outside world, and it feels like it might be overkill here. However, I think we should validate _somewhere_. I'm not sure if this needs to be part of this task, last week @gareth-allan mentioned to me that we've got some other task for sorting out the HTTP interface, so I'd imagine we'd do HTTP request validation at that time?

Philip-Woods replied:
Fair point, we'll have to remember to do this somewhere else that's appropriate

ismail-s replied:
I've mentioned this to Colin to put into the relevant story (as more of a general thing, not just for message id).

Philip-Woods commented on mhs/mhs/outbound/request/synchronous/handler.py:
```diff
@@ -28,15 +28,16 @@ def initialize(self, workflow: sync_async_workflow.SyncAsyncWorkflow,
 
-    async def post(self):
+    async def post(self, interaction_name):
         logging.debug("Client POST received: %s", self.request)
```
(Optional as maybe it's worth refactoring it all at a later date) I know you didn't write this but we maybe we should try to update old logging statements as we go?

ismail-s replied:
I agree (and @gareth-allan mentioned this to me as a general point as well).

ismail-s committed:
RT-93 Don't import classes, import typing classes

ismail-s committed:
RT-93 change exception name

ismail-s committed:
RT-93 include key in log message

ismail-s committed:
RT-93 use common logger

ismail-s committed:
RT-93 remove uused import

ismail-s committed:
RT-93 make a copy of interaction details

This prevents users of ConfigurationManager modifying the config and
causing conflicts with each other.

ismail-s committed:
RT-93 make process keys more readable
